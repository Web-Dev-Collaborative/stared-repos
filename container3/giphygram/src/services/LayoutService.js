// Set of screen sizes that app is supporting.
export const SCREEN_SIZES = {
  xl: { id: 'xl', mediaQuery: '(min-width: 1200px)' },
  lg: { id: 'lg', mediaQuery: '(max-width: 1200px) and (min-width: 960px)' },
  md: { id: 'md', mediaQuery: '(max-width: 950px) and (min-width: 720px)' },
  sm: { id: 'sm', mediaQuery: '(max-width: 720px) and (min-width: 540px)' },
  xs: { id: 'xs', mediaQuery: '(max-width: 540px)' },
};

// Supported number of columns in 12-column templates.
export const SUPPORTED_COLUMNS_NUMS = [1, 2, 3, 4, 6, 12];

export const DEFAULT_COLUMNS_NUM = 3;

export const TOTAL_NUMBER_OF_LAYOUT_COLUMNS = 12;

export class LayoutService {
  /**
   * Get default number of columns for 12-column template depending on screen size.
   * @param {string} screenSize
   * @return {number}
   */
  static screenSizeToColumns(screenSize) {
    switch (screenSize) {
      case SCREEN_SIZES.xs.id:
        return 1;

      case SCREEN_SIZES.sm.id:
      case SCREEN_SIZES.md.id:
        return 3;

      case SCREEN_SIZES.lg.id:
      case SCREEN_SIZES.xl.id:
        return 4;

      default:
        return DEFAULT_COLUMNS_NUM;
    }
  }

  /**
   * Chooses the number of layout columns from manual and automatic one.
   * @param {?number} responsiveColumnsNum - number of layout columns according to window size
   * @param {?number} manualColumnsNum - number of layout columns according to user selection.
   * @return {number}
   */
  static choseColumnsNum(responsiveColumnsNum, manualColumnsNum) {
    // Manual column selection has higher priority than automatic columns number
    // calculated based on window size.
    return manualColumnsNum || responsiveColumnsNum;
  }
}
