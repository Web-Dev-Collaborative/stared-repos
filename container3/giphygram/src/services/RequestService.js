import axios from 'axios';
import { REQUEST_TIMEOUT } from '../config';

// RequestService is created as a wrapper on top of axios. This is done in order to be able to
// switch to new http-requests library if needed. Utilizing interfaces might be a good option here.
export class RequestService {
  // Perform GET request with parameters.
  static get(url, params = {}) {
    return axios.get(
      url,
      { ...this.getDefaultAxiosConfig(), params },
    );
  }

  // Generate default Axios configuration for requests.
  // We may put here default timeouts etc.
  static getDefaultAxiosConfig() {
    return {
      timeout: REQUEST_TIMEOUT,
    };
  }
}
