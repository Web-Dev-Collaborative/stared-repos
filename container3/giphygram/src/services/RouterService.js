const SEARCH_QUERY_PARAM = 'query';

export class RouterService {
  constructor(history, location) {
    this.history = history;
    this.location = location;
  }

  pushSearchQuery(query = '') {
    this.history.push({
      search: query ? `?${SEARCH_QUERY_PARAM}=${query}` : '',
    });
  }

  getSearchQuery() {
    const searchParams = new URLSearchParams(this.location.search);
    return searchParams.get(SEARCH_QUERY_PARAM);
  }
}
