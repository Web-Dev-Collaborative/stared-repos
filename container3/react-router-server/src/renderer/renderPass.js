import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import AsyncRenderer from '../components/AsyncRenderer';
import removeDuplicateModules from '../utils/removeDuplicateModules';

const renderPass = (context, element, staticMarkup = false) => {
  context.callback = () => {
    if (context.finishedLoadingModules && !context.statesRenderPass) {
      context.statesRenderPass = true;
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    } else if (context.finishedLoadingModules && context.statesRenderPass || !context.hasModules) {
      context.renderResult = renderPass(context, element, staticMarkup);
      if (context.fetchingStates <= 0 && context.modulesLoading <= 0) {
        context.resolve({
          html: context.renderResult,
          state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
          modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
        });
      }
    }
  };

  let component = (
    <AsyncRenderer context={context}>
      {element}
    </AsyncRenderer>
  );

  let result;
  try {
    if (staticMarkup) {
      result = renderToStaticMarkup(component);
    } else {
      result = renderToString(component);
    }
  } catch (e) {
    return context.reject(e);
  }

  if (!context.hasModules && !context.hasStates) {
    context.resolve({
      html: result,
      state: context.fetchStateResults === undefined ? null : context.fetchStateResults,
      modules: context.modules === undefined ? null : removeDuplicateModules(context.modules)
    });
  }
  return result;
};

export default renderPass;
