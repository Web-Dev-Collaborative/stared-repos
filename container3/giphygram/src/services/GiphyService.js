import { RequestService } from './RequestService';
import { GIPHY_API_HOST, GIPHY_API_KEY } from '../config';
import { defaultSearchParams } from '../reducers/searchParamsReducer';

// GiphyService is responsible for all integrations with Giphy API.
export class GiphyService {
  // Search for GIFs by specific search query.
  static search({
    query = defaultSearchParams.query,
    limit = defaultSearchParams.limit,
    offset = defaultSearchParams.offset,
    rating = defaultSearchParams.rating,
    lang = defaultSearchParams.lang,
  } = {}) {
    const searchParams = {
      api_key: GIPHY_API_KEY,
      q: query,
      limit,
      offset,
      rating,
      lang,
    };

    return RequestService.get(
      `${GIPHY_API_HOST}/v1/gifs/search`,
      searchParams,
    );
  }
}
