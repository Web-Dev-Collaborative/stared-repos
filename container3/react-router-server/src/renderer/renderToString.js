import renderPass from './renderPass';
import { clear } from '../module/cache';

export default (element) => {
  return new Promise((resolve, reject) => {
    clear();
    const context = {
      resolve, reject,
      modulesLoading: 0,
      fetchingStates: 0
    };
    renderPass(context, element);
  });
};
