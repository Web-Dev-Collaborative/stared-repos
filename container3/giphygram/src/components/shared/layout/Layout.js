import React from 'react';
import PropTypes from 'prop-types';
import { SCREEN_SIZES } from '../../../services/LayoutService';

// Closure callback that fires when media has been changed.
const onMediaQueryChangeCallback = (screenSize, callback) => (e) => {
  if (e.matches) {
    callback(screenSize.id);
  }
};

export class Layout extends React.Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.node),
      PropTypes.node,
    ]).isRequired,
    onMediaQueryChange: PropTypes.func,
    onMediaQueryInit: PropTypes.func,
  };

  static defaultProps = {
    onMediaQueryChange: () => {},
    onMediaQueryInit: () => {},
  };

  constructor(props) {
    super(props);

    // List of all listeners to media query changes.
    this.mediaQueries = {};
  }

  componentDidMount() {
    const { onMediaQueryChange, onMediaQueryInit } = this.props;
    let initialScreenSize = null;

    Object
      .values(SCREEN_SIZES)
      .forEach((screenSize) => {
        // Try to match current media query.
        this.mediaQueries[screenSize.id] = window.matchMedia(screenSize.mediaQuery);
        if (this.mediaQueries[screenSize.id].matches) {
          initialScreenSize = screenSize;
        }

        // Subscribe to media query changes.
        this.mediaQueries[screenSize.id].addListener(
          onMediaQueryChangeCallback(screenSize, onMediaQueryChange),
        );
      });

    // Setup initial media query match.
    onMediaQueryInit(initialScreenSize.id);
  }

  // @TODO: Remove media query listeners on component unmount.
  // componentWillUnmount() {
  //   Object.values(this.mediaQueries).forEach(mediaQuery => mediaQuery.removeListener());
  // }

  render() {
    const { children } = this.props;

    return (
      <div className="container">
        <div className="row">
          <div className="col">
            {children}
          </div>
        </div>
      </div>
    );
  }
}
