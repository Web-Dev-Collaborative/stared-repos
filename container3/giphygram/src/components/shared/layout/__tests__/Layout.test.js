import React from 'react';
import renderer from 'react-test-renderer';
import { Layout } from '../Layout';
import { SCREEN_SIZES } from '../../../../services/LayoutService';

describe('Layout', () => {
  it('should render one child component inside the layout', () => {
    const tree = renderer
      .create((
        <Layout>
          <div>One and only child</div>
        </Layout>
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render more than one child component inside the layout', () => {
    const tree = renderer
      .create((
        <Layout>
          <div>First child</div>
          <div>Second child</div>
        </Layout>
      ))
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should fire onMediaQueryInit callback', () => {
    const onMediaQueryInit = jest.fn();

    renderer.create((
      <Layout onMediaQueryInit={onMediaQueryInit}>
        <div>One and only child</div>
      </Layout>
    ));

    expect(onMediaQueryInit).toHaveBeenCalledTimes(1);
    expect(onMediaQueryInit).toHaveBeenCalledWith('xs');
  });

  it('should fire onMediaQueryChange callback when media query matches', () => {
    // Mock window.matchMedia() method.
    // Normally when we add listener it doesn't fire up.
    // But here we're fire the listener just after the subscription
    // for testing purpose only.
    global.matchMedia = jest.fn().mockImplementation(query => ({
      matches: true,
      media: query,
      onchange: null,
      addListener: callback => callback({ matches: true }),
      removeListener: jest.fn(),
    }));

    const onMediaQueryChange = jest.fn();
    const onMediaQueryInit = jest.fn();

    renderer.create((
      <Layout
        onMediaQueryChange={onMediaQueryChange}
        onMediaQueryInit={onMediaQueryInit}
      >
        <div>One and only child</div>
      </Layout>
    ));

    expect(onMediaQueryInit).toHaveBeenCalledTimes(1);
    expect(onMediaQueryInit).toHaveBeenCalledWith('xs');

    expect(onMediaQueryChange).toHaveBeenCalledTimes(Object.keys(SCREEN_SIZES).length);
    expect(onMediaQueryChange.mock.calls[0][0]).toEqual(SCREEN_SIZES.xl.id);
    expect(onMediaQueryChange.mock.calls[1][0]).toEqual(SCREEN_SIZES.lg.id);
    expect(onMediaQueryChange.mock.calls[2][0]).toEqual(SCREEN_SIZES.md.id);
    expect(onMediaQueryChange.mock.calls[3][0]).toEqual(SCREEN_SIZES.sm.id);
    expect(onMediaQueryChange.mock.calls[4][0]).toEqual(SCREEN_SIZES.xs.id);
  });

  it('should not fire onMediaQueryChange callback when media query does not match', () => {
    // Mock window.matchMedia() method.
    // Normally when we add listener it doesn't fire up.
    // But here we're fire the listener just after the subscription
    // for testing purpose only.
    global.matchMedia = jest.fn().mockImplementation(query => ({
      matches: true,
      media: query,
      onchange: null,
      addListener: callback => callback({ matches: false }),
      removeListener: jest.fn(),
    }));

    const onMediaQueryChange = jest.fn();
    const onMediaQueryInit = jest.fn();

    renderer.create((
      <Layout
        onMediaQueryChange={onMediaQueryChange}
        onMediaQueryInit={onMediaQueryInit}
      >
        <div>One and only child</div>
      </Layout>
    ));

    expect(onMediaQueryInit).toHaveBeenCalledTimes(1);
    expect(onMediaQueryInit).toHaveBeenCalledWith('xs');

    expect(onMediaQueryChange).not.toHaveBeenCalled();
  });
});
