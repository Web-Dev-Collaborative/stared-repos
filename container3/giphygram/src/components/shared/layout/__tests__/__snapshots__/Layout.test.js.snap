// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render more than one child component inside the layout 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div>
        First child
      </div>
      <div>
        Second child
      </div>
    </div>
  </div>
</div>
`;

exports[`Layout should render one child component inside the layout 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <div>
        One and only child
      </div>
    </div>
  </div>
</div>
`;
